syntax/sd.vim

```diff
@@ -50,54 +71,111 @@ syn keyword ReturnKeyword かえる
-" bol or whitespace, number, followed by a comma, whitespace, or a comment
-syn match NumberMatch /\v(^|[ 　])-?(\d+\.\d+|\d+)((から|まで|で|と|に|へ|を)[ 　]|[?？])@=/
-" bol or whitespace, number, followed by a particle or question mark
-
-syn match CommentMatch /\v[(（].*$/ contains=TodoKeyword
-
-syn match CompParamMatch /\v[^ 　]{-}((が|\?|？|と|より|以上|以下)([ 　][ 　]{-})@=)@=/ contained
-syn match CompParticleMatch /\v([^ 　]{-})@<=(が|\?|？|と|より|以上|以下)([ 　][ 　]{-})@=/ contained
-syn match CompFuncCallParamMatch /\v(が[ 　])@<![^ 　]{-}((から|まで|で|と|に|へ|を)([ 　][ 　]{-})@=)@=/ contained
+let specialGroup  = '(それ|あれ)'
+let boolGroup     = '(真|肯定|はい|正|偽|否定|いいえ)'
+let nullGroup     = '(無|無い|無し|ヌル)'
+let arrayGroup    = '(配列)' " TODO: (v1.1.0) add 連想配列
+let particleGroup = '(から|まで|で|と|に|へ|を)'
+let ifElseIfGroup = '(もし|もしくは|または)'
+let elseGroup     = '(それ以外|違えば|ちがえば)'
+let comp12Group   = '(が|\?|？|と|より|以上|以下)'
+let comp3Group    = '(ならば|でなければ|(等し(くな)?|ひとし(くな)?|小さ|ちいさ|短|みじか|低|ひく|少な|すくな|大き|おおき|長|なが|高|たか|多|おお)ければ)'
+
+let whitespaceRegion    = '[ 　]'
+let notWhitespaceRegion = '[^ 　]'
+let commaRegion         = '[,、]'
+let separatorRegion     = '[ ,　、]'
+let notSeparatorRegion  = '[^ ,　、]'
+let commentStartRegion  = '[(（]'
+let questionRegion      = '[?？]'
+
+let number = '-?(\d+\.\d+|\d+)'
+let bol    = '^' . whitespaceRegion . '*'
+let eol    = whitespaceRegion . '*(' . commentStartRegion . '.*)?$'
 
-syn match VarDefMatch /\v(^[ 　]*[^ ,　、]+)@<=は([ 　])@=/
-
-syn match FuncDefMatch /\v^.*[^ ,　、]+とは/
-        \ contains=FuncDefLeadingWhitespaceMatch,FuncDefParamMatch,FuncDefParticleMatch,FuncDefNameMatch
-
-syn match FuncDefLeadingWhitespaceMatch /\v^[ 　]*/
-        \ contained nextgroup=FuncDefParamMatch
-syn match FuncDefParamMatch /\v([ 　]*)@<=[^ ,　、][^ ,　、]{-}((から|まで|で|と|に|へ|を)[ 　])@=/
-        \ contained
-" non whitespace, particle (not preceding は), followed by whitespace
-syn match FuncDefParticleMatch /\v([^ 　])@<=(から|まで|で|と|に|へ|を)(は$)@!([ 　][ 　]{-})@=/
-        \ contained
-" zero or more whitespace, a name, followed by とは and zero or more whitespace
-syn match FuncDefNameMatch /\v([ 　]*)@<=[^ ,　、]+(とは[ 　]*((\(|（).*)?$)@=/
-        \ contained
-
-" non whitespace, particle, followed by whitespace
-syn match FuncCallParticleMatch /\v([^ 　])@<=(から|まで|で|と|に|へ|を)[ 　]@=/
-syn match FuncCallGlobalSpecialMatch /\v(それ|あれ)((から|まで|で|と|に|へ|を)[ 　])@=/
-
-" a name, followed by a space and というものは
-syn match ClassDefMatch /^\v[^ ,　、]*[ 　]+と(い|言)う(もの|物)は/
-        \ contains=ClassDefLeadingWhitespaceMatch
-
-syn match ClassDefLeadingWhitespaceMatch /\v^[ 　]*/
-        \ contained nextgroup=ClassDefNameMatch
-syn match ClassDefNameMatch /\v([ 　]*)@<=[^ ,　、]+([ 　]+と(い|言)う(もの|物)は)@=/
-        \ contained
+"-------------------------------------------------------------------------------
+" Matches
+"-------------------------------------------------------------------------------
+exe 'syn match SpecialKeyword /\v^' . whitespaceRegion . '*' . specialGroup . '(は)@=/'
```

Comment:
Although a tiny bit ugly, if I replace the `syn match ...` commands with dynamically-built `exe` (basically `eval`), it allows me to include variables, which makes a lot of the regex wayyy easier to read.